donkeycar/parts/tub_v2.py

```diff
@@ -0,0 +1,112 @@
+import json
+import os
+import time
+
+import numpy as np
+from PIL import Image
+
+from donkeycar.parts.datastore_v2 import Manifest, ManifestIterator
+
+
+class Tub(object):
+    '''
+    A datastore to store sensor data in a key, value format. \n
+    Accepts str, int, float, image_array, image, and array data types.
+    '''
+    def __init__(self, base_path, inputs=[], types=[], metadata=[], max_catalog_len=1000):
+        self.base_path = base_path
+        self.images_base_path = os.path.join(self.base_path, Tub.images())
+        self.inputs = inputs
+        self.types = types
+        self.metadata = metadata
+        self.manifest = Manifest(base_path, inputs=inputs, types=types, metadata=metadata, max_len=max_catalog_len)
+        self.input_types = dict(zip(self.inputs, self.types))
+        # Create images folder if necessary
+        if not os.path.exists(self.images_base_path):
+            os.makedirs(self.images_base_path, exist_ok=True)
+
+    def write_record(self, record):
+        '''
+        Can handle various data types including images.
+        '''
+        contents = dict()
+        for key, value in record.items():
+            if value is None:
+                continue
+            elif not key in self.input_types:
+                continue
+            else:
+                input_type = self.input_types[key]
+                if input_type == 'float':
+                    # Handle np.float() types gracefully
+                    contents[key] = float(value)
+                elif input_type == 'str':
+                    contents[key] = value
+                elif input_type == 'int':
+                    contents[key] = int(value)
+                elif input_type == 'boolean':
+                    contents[key] = bool(value)
+                elif input_type == 'list' or input_type == 'vector':
+                    contents[key] = list(value)
+                elif input_type == 'image_array':
+                    # Handle image array
+                    image = Image.fromarray(np.uint8(value))
+                    name = Tub._image_file_name(self.manifest.current_index, key)
+                    image_path = os.path.join(self.images_base_path, name)
+                    image.save(image_path)
+                    contents[key] = name
+
+        # Private properties
+        contents['_timestamp'] = int((time.time() / 1000) * 1000)
```

Comment:
Can you please explain this?
Division and then multiplication?

Reply:
`time.time()` returns a floating point timestamp. Dividing and multiplying by a `1000` gives us the timestamp in milliseconds.

Reply:
Maybe `time.monotonic()` to avoid issues of clock skews on some systems?
(I've seen really strange things....)

Or `time.monotonic_ns()` which returns int but it requires python 3.7

Reply:
This is effectively just a rounding, but because divisions are expensive, not be best way to do it, `round` would be faster here.